internal/hooks/v0hooks/manager.go

```diff
@@ -188,47 +97,116 @@ func (o *Manager) runHook(
 	return nil
 }
 
-func checkError(
-	hookOutput HookOutput,
+func configByName(
+	cfg *conf.HookConfiguration,
+	name Name,
+) (*conf.ExtensibilityPointConfiguration, bool) {
+	switch name {
+	case SendSMS:
+		return &cfg.SendSMS, true
+	case SendEmail:
+		return &cfg.SendEmail, true
+	case CustomizeAccessToken:
+		return &cfg.CustomAccessToken, true
+	case MFAVerification:
+		return &cfg.MFAVerificationAttempt, true
+	case PasswordVerification:
+		return &cfg.PasswordVerificationAttempt, true
+	case BeforeUserCreated:
+		return &cfg.BeforeUserCreated, true
+	case AfterUserCreated:
+		return &cfg.AfterUserCreated, true
+	default:
+		return nil, false
+	}
+}
+
+func (o *Manager) Enabled(name Name) bool {
+	if cfg, ok := configByName(&o.config.Hook, name); ok {
+		return cfg.Enabled
+	}
+	return false
+}
+
+func (o *Manager) BeforeUserCreated(
+	ctx context.Context,
+	tx *storage.Connection,
+	req *BeforeUserCreatedRequest,
+	res *BeforeUserCreatedResponse,
 ) error {
-	if hookOutput.IsError() {
-		he := hookOutput.GetHookError()
-		httpCode := he.HTTPCode
+	return o.dr.Dispatch(ctx, &o.config.Hook.BeforeUserCreated, tx, req, res)
+}
 
-		if httpCode == 0 {
-			httpCode = http.StatusInternalServerError
-		}
+func (o *Manager) AfterUserCreated(
+	ctx context.Context,
+	tx *storage.Connection,
+	req *AfterUserCreatedRequest,
+	res *AfterUserCreatedResponse,
+) error {
+	return o.dr.Dispatch(ctx, &o.config.Hook.AfterUserCreated, tx, req, res)
+}
 
-		httpError := &apierrors.HTTPError{
-			HTTPStatus: httpCode,
-			Message:    he.Message,
-		}
-		return httpError.WithInternalError(&he)
-	}
-	return nil
+func (o *Manager) InvokeHook(
+	conn *storage.Connection,
+	r *http.Request,
+	input, output any,
+) error {
+	return o.invokeHook(conn, r, input, output)
 }
 
-func validateTokenClaims(outputClaims map[string]interface{}) error {
-	schemaLoader := gojsonschema.NewStringLoader(MinimumViableTokenSchema)
+// invokeHook invokes the hook code. conn can be nil, in which case a new
+// transaction is opened. If calling invokeHook within a transaction, always
+// pass the current transaction, as pool-exhaustion deadlocks are very easy to
+// trigger.
+func (o *Manager) invokeHook(
+	conn *storage.Connection,
+	r *http.Request,
+	input, output any,
+) error {
```

Comment:
I would much rather see the before and access hooks listed here. Easy to debug, easy to spot, no unnecessary indirection.

Reply:
I actually planned to go the other direction and eventually make sure to add methods with strong typing to reduce the need for more control flow and type assertions used in the legacy hooks.